### PR TITLE
AI service rework

### PR DESCRIPTION
**Pull request by xunkar**
This PR reworks the AI service with a large increase in performance, new features, and fixes. Here is an overview of what the PR entails:

– The AI service workflow has been redesigned to be compatible with threading thanks to the use of `retro_task`, removing any lag in emulation when invoking the service and when displaying the results.
– The performance increase makes it possible to run the service in an automated way. The automatic mode can only be enabled by the AI server if it supports it however.
– Automatic translation features a duplicate request checker, that avoids spamming the server if no new data needs to be translated.
– Automatic translation is constrained by an automatic poll delay that can be configured by the user. Because automatic mode can have a visible impact on the CPU (due to the encoding/decoding of frames) this delay gives some amount of control to the user to choose between reactivity and performance.
– The AI service now includes a text mode that displays a returned string in the manner of subtitles. The placement of the subtitle can be hinted at by the server, and controlled and overriden by the user.
– The AI service now includes a text + narrator and image + narrator, allowing to combine both visual results and TTS results.
– Improved user feedback and error handling throughout.
– Full documentation is included in `accessibility.h` for anyone to implement their own AI server in the future.
– Fixes a bug where the accessibility narrator would not fallback to a different method if NVDA failed. (This method should properly be reworked in the future though, but I had concerns doing it during this project.)

I've kept the legacy code for `translation_response_image_direct` that was already there (simply encapsulated it in its own method), but I have been unable to make it work in any shape or form. This method is supposed to serve as an alternative to widgets to draw returned images directly onto the frame buffer, but nothing happens on my end. It would be wiser to disable it altogether for the moment, and warn the user that image mode requires widgets to be enabled, but perhaps the issue lies with my setup.

Image overlay mode was unfortunately incompatible with non-core provided aspect ratios. I have found how to fix this issue, but it creates a different kind of problem with vulkan, where requesting the size of the viewport somehow modifies the result, messing up with all widgets. Until a solution can be provided, I've added viewport data to the request so that AI servers can create an image of the correct size.

Here's a short [video](https://youtu.be/7oBEoivqBm4) demonstrating some of the changes. This automatic mode is made possible by a custom local service I'm working on.

### CODE REVIEW

**Review by BarryJRowe — COMMENTED** [review]
I need to do some testing on the pr locally. I will comment again once I have.

File: gfx/gfx_widgets.c [hunk 1]
```diff
@@ -1563,63 +1620,46 @@ void gfx_widgets_frame(void *data)
          if (dispctx->blend_end)
             dispctx->blend_end(userdata);
       }
-
-      /* top line */
-      gfx_display_draw_quad(
```
[the remark this answers is not in the document]
Comment (xunkar) [reply gfx/gfx_widgets.c:1568]: This is no longer necessary as now the service warns the user. A message (`runloop_msg_queue_push`) is pushed in all of the following cases:

– The server returns an empty string or empty picture (manual requests only)
– The server enables or disables automatic translation
– Automatic translation is disabled by the user (by using the AI toggle bind or pulling the menu)
– The server returns an error
– The server returns a malformed JSON
– The service encounters an internal error

**Comment by LibretroAdmin** [comment]
Hi, how are things going with this PR? Has @BarryJRowe been able to talk with you about some things?

**Comment by xunkar** [comment]
There's currently an issue that only appears with OpenGL-based drivers (`gl` and `glcore`) that I've been unable to track down. Somehow the overlay used by the AI service is mysteriously released, or the pointer becomes stale. There's no such issue with vulkan or d3d, and I haven't changed anything in the way the overlay texture is loaded or drawn, so I'm at a loss here. I hope to have enough time soon to delve deeper into this issue, but I'm not sure I can do anything about it myself actually.

**Comment by LibretroAdmin** [comment]
Hi @neil4 , would you be able to help with this issue?

**Comment by neil4** [comment]
I can take a look this weekend (out of town right now)

**Comment by neil4** [comment]
I don't think you can load a texture from a separate thread in OpenGL, at least not without making the context current for that thread. (I'm not a graphics guru, so take that with a grain of salt)

I was able to see it work with glcore by moving the `translation_response_image_widget` call from `translation_response_image_hndl` to `translation_response_image_cb` so that texture loading (`gfx_widgets_ai_service_overlay_load`) is on the main thread. That change alone doesn't cover all cases of texture _unloading_, but you get the idea - texture loading and unloading have to occur on the main thread.

**Comment by xunkar** [comment]
Hm, this might prove problematic. I'll try to defer all those calls to the main thread, but ultimately we might be back with a service that overtasks the main thread and cauises lag. I'll see what can be done, thank you for that.

**Comment by xunkar** [comment]
Moved the texture loading call back to the main thread, which indeed fixes the issue. Performance doesn't seem impacted. I think we should be good. Also, unloading is only ever invoked on the main thread. (Back to you, @BarryJRowe)

**Comment by Cpod12** [comment]
Are you still working on this @xunkar I'd love to help in any way I can if you are open to it. I think this is super cool what you're working on!

**Comment by xunkar** [comment]
@Cpod12 thank you. The changes were done and ready to be merged since my last comment, there has been no updates from the team. Now I see there are conflicting files, but I'm not particularly comfortable with github and I'm not sure what the conflicts are or how to solve them. I'm not seeing anything conflicting when I click on "Resolve conflicts" 🤔

**Comment by hizzlekizzle** [comment]
We ran into this with another PR recently where no conflicts were visible to the submitter but github wouldn't let it merge cleanly. In their case, the solution was to close their original PR and just open a new one and merge it immediately.

If this one is ready to go and approved by @LibretroAdmin , that may be the most expedient solution here, as well.

**Comment by Cpod12** [comment]
@xunkar just lmk if you need help with this or something similar in the future! Hopefully, you're able to merge it soon :)

**Comment by Cpod12** [comment]
@xunkar maybe it has to do with commits that were added after the pull request? want me to take a look and see if I can fix it?

**Comment by Cpod12** [comment]
I looked and don't think I can resolve all the conflicts on my own. I don't understand what's going on. @xunkar would you be fine fetching those changes and merging that stuff manually? Once you do that and push it it should be good to go

**Comment by Cpod12** [comment]
I almost got it working but when I try to compile it I get an error that says "undefined reference to `video_driver_cached_frame_get'". Is this a function that no longer exists? What can I use instead to replace it?

**Comment by Cpod12** [comment]
Nvm I figured it out! Got fixed the merge conflicts. :) What should I do now? Would I make my own pull request or what?

**Comment by hizzlekizzle** [comment]
Sure, you can open another one and we'll see if it's merge-able

**Pull request closed without merging**